clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManager.java

```diff
@@ -59,4 +59,21 @@ public interface MembershipManager {
      * current assignment.
      */
     void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assignment);
+
+    /**
+     * Transition the member to the FENCED state.  This is only invoked when the heartbeat returns a
+     * FENCED_MEMBER_EPOCH or UNKNOWN_MEMBER_ID error code.
+     */
+    void fenceMember();
+
+    /**
+     * Transition the member to the FAILED state.  This is invoked when the heartbeat returns an UNRELEASED_MEMBER_ID
```

Comment:
This will be invoked on any non-retriable error I expect (not only the UNRELEASED_MEMBER_ID)